hydra/utils.py

```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
         return None
 
-    if isinstance(config, TargetConf) and config._target_ == "???":
+    # TargetConf edge case
+    if isinstance(config, TargetConf) and config.is_missing():
         # Specific check to give a good warning about failure to annotate _target_ as a string.
         raise InstantiationException(
             f"Missing value for {type(config).__name__}._target_. Check that it's properly annotated and overridden."
             f"\nA common problem is forgetting to annotate _target_ as a string : '_target_: str = ...'"
         )
 
-    if not (
-        isinstance(config, dict)
-        or OmegaConf.is_config(config)
-        or is_structured_config(config)
-    ):
-        raise HydraException(f"Unsupported config type : {type(config).__name__}")
+    # Structured Config always converted first to OmegaConf
+    if is_structured_config(config):
+        config = OmegaConf.structured(config, flags={"allow_objects": True})
 
+    # Make shallow copy if config is a standard dictionary
     if isinstance(config, dict):
-        configc = config.copy()
-        _convert_container_targets_to_strings(configc)
-        config = configc
-
-    kwargsc = kwargs.copy()
-    _convert_container_targets_to_strings(kwargsc)
-    kwargs = kwargsc
+        config = config.copy()
 
-    # make a copy to ensure we do not change the provided object
-    config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+    # If OmegaConf, copy, make it mutable and set parent for live interpolation
     if OmegaConf.is_config(config):
+        config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+        OmegaConf.set_readonly(config_copy, False)
+        OmegaConf.set_struct(config_copy, False)
         config_copy._set_parent(config._get_parent())
-    config = config_copy
-    assert OmegaConf.is_config(config)
-    OmegaConf.set_readonly(config, False)
-    OmegaConf.set_struct(config, False)
+        config_copy._set_flag("allow_objects", True)
+        config = config_copy
 
-    target = _get_target_type(config, kwargs)
-    try:
-        config._set_flag("allow_objects", True)
-        final_kwargs = _get_kwargs(config, **kwargs)
-        convert = _pop_convert_mode(final_kwargs)
-        if convert == ConvertMode.PARTIAL:
-            final_kwargs = OmegaConf.to_container(
-                final_kwargs, resolve=True, exclude_structured_configs=True
+    # Finalize config (convert targets to strings, merge with kwargs)
+    # TODO: kwargs are used as a config now. If it contains dataclass
+    # objects, they will be interpreted as config nodes in the
+    # instantiation result. We might want to change that behavior.
+    _convert_container_targets_to_strings(kwargs)
+    _convert_container_targets_to_strings(config)
+    config = _merge(config, kwargs) if kwargs else config
+
+    # Override parent modes from config if specified
+    # Cannot use pop with default because of OmegaConf typed config
```

Comment:
Not sure I understand the statement.
if DictConfig.pop() behaves differently than native dict pop(), please file an issue with a minimal repro against OmegaConf.

Reply:
I left a comment on an opened issue https://github.com/omry/omegaconf/issues/425.
I'm not sure this is really a bug, more like some type checking functionality of OmegaConf that forces us to "look before trying"

Reply:
I opened a separate issue at your request here https://github.com/omry/omegaconf/issues/471